x/btcstkconsumer/keeper/msg_server.go

```diff
@@ -19,19 +21,61 @@ func NewMsgServerImpl(keeper Keeper) types.MsgServer {
 
 var _ types.MsgServer = msgServer{}
 
-// RegisterConsumer registers a CZ consumer
-func (ms msgServer) RegisterConsumer(goCtx context.Context, req *types.MsgRegisterConsumer) (*types.MsgRegisterConsumerResponse, error) {
-	consumerRegister := &types.ConsumerRegister{
-		ConsumerId:          req.ConsumerId,
-		ConsumerName:        req.ConsumerName,
-		ConsumerDescription: req.ConsumerDescription,
+// RegisterConsumer registers a consumer
+func (ms msgServer) RegisterConsumer(ctx context.Context, req *types.MsgRegisterConsumer) (*types.MsgRegisterConsumerResponse, error) {
+	// if the permissioned integration is enabled and the signer is not the authority
+	// then this is not an authorised registration request, reject
+	if ms.GetParams(ctx).PermissionedIntegration && ms.authority != req.Signer {
+		return nil, errorsmod.Wrapf(govtypes.ErrInvalidSigner, "invalid authority; expected %s, got %s", ms.authority, req.Signer)
 	}
 
-	ctx := sdk.UnwrapSDKContext(goCtx)
-	err := ms.Keeper.RegisterConsumer(ctx, consumerRegister)
-	if err != nil {
+	if err := req.ValidateBasic(); err != nil {
 		return nil, err
 	}
 
+	if len(req.EthL2FinalityContractAddress) > 0 {
+		// this is a ETH L2 consumer
+
+		// ensure the ETH L2 finality contract exists
+		contractAddr, err := sdk.AccAddressFromBech32(req.EthL2FinalityContractAddress)
+		if err != nil {
+			return nil, types.ErrInvalidETHL2ConsumerRequest.Wrapf("%s", err.Error())
+		}
+		contractInfo := ms.wasmKeeper.GetContractInfo(ctx, contractAddr)
+		if contractInfo == nil {
+			return nil, types.ErrInvalidETHL2ConsumerRequest.Wrapf("ETH L2 finality contract does not exist")
+		}
+
+		// all good, register this ETH L2 consumer
+		consumerRegister := types.NewETHL2ConsumerRegister(
+			req.ConsumerId,
+			req.ConsumerName,
+			req.ConsumerDescription,
+			req.EthL2FinalityContractAddress,
+		)
+		if err := ms.Keeper.RegisterConsumer(ctx, consumerRegister); err != nil {
+			return nil, err
+		}
+	} else {
+		// this is a Cosmos consumer
+
+		// ensure the IBC light client exists
+		sdkCtx := sdk.UnwrapSDKContext(ctx)
+		_, exist := ms.clientKeeper.GetClientState(sdkCtx, req.ConsumerId)
```

Comment:
Wouldn't this require that the IBC connection is established **before** registering the consumer? What is the right / required sequence for Consumer registration?

Reply:
Not really, here we are retrieving IBC *light client* but not connection or channel.

So the user story now becomes: creating an IBC lgiht client for consumer -> registering consumer using this client ID -> passing gov prop if permissioned integration is enforced -> creating IBC connection/channel

This is the approach followed by ICS, as discussed in the RFC/ADR